bpf: nat: skip snat when using iptables masquerade

gyutaeb opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

When using iptables masquerade, the source address is already nat-ed by iptables. So, we should skip snat when using iptables masquerade. If we don't skip snat, the source address is nat-ed twice. It also unnecessarily occupies entry in the SNAT_MAPPING map. And if it exceeds SNAT_COLLISION_RETRIES, it leads to connection failures.

Fixes: #36572

julianwiedmann reviewed (changes requested):
>It is no longer correct to check the NAT_MIN_EGRESS condition. This is because target->from_local_endpoint is now being checked. In the case of !target->from_local_endpoint, source NAT is not required,

I don't follow. Can you please elaborate *much* more on the suggested change. **Why** is source NAT not required in the case you're describing? Also please add this in-depth explanation to the patch description, not just the PR.

gyutaeb committed:
bpf: nat: skip snat when using iptables masquerade

When using iptables masquerade, the source address is already nat-ed by
iptables. So, we should skip snat when using iptables masquerade. If
we don't skip snat, the source address is nat-ed twice. It also
unnecessarily occupies entry in the SNAT_MAPPING map. And if it exceeds
SNAT_COLLISION_RETRIES, it leads to connection failures.

Fixes: #36572

Signed-off-by: Gyutae Bae <gyu.8ae@gmail.com>

julianwiedmann reviewed (changes requested):
I don't see how this can be correct for the case of using BPF Nodeport without BPF masquerade. Please elaborate a lot more.

We're **intentionally** tracking connections in the SNAT engine for this case, and applying SNAT as needed to avoid conflicts between [LB-to-remote-backend connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport.h#L2453) and [host-originating / iptables-masqueraded connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport_egress.h#L347).
